src/coord/src/persistence.rs

```diff
@@ -112,39 +112,33 @@ impl Source {
             // Keep only the minimum timestamp we received for every offset
             partition.pending.dedup_by_key(|x| x.offset);
 
-            let mut prev = partition.last_persisted_offset;
+            let mut most_recent_offset = partition.last_persisted_offset;
             let mut prefix_length = 0;
             let mut prefix_start_offset = None;
 
+            // Find the longest unbroken prefix where each subsequent record refers to the
```

Comment:
So I am having a hard time figuring out what happens if we have two dataflows reading the same source at the same time, while there is compaction also happening in parallel so for example

if the first dataflow observes (predecessor, offset) like

(None, 0), (0, 1), (1, 2)

and in parallel the second dataflow observes

(None, 0), (0, 2) /// so the 1 got compacted away.


my understanding is that all of these would be in the same array, and they would be sorted with the default comparator which iiuc compares things in field order, so assuming that everything else around the key, value and timestamp at a given offset stays the same means we would end up with:

(None, 0), (None, 0), (0, 1), (0, 2) (1, 2) // because we now sort by predecessor first, then offset, then timestamp

and then we would dedup by offset, which would leave us with

(None, 0), (0, 1), (0, 2) // which would fail to persist the (0, 2) record

does that sound right / make sense?

Reply:
I don't immediately have a thought for how we should be doing the sorting and deduplication in this case - my gut says that we should keep the largest gaps between offsets because we know that the data in the offsets that were skipped over won't be useful anyway

which i think translates to sort by (offset, predecessor) and then keep the record with the min predecessor among all records that are at `offset`?

but we should think this through a bit more and I also think we should extract this logic that gets a prefix out of a array of records into a function so that we can do a cargo test for it maybe?